<a href="https://colab.research.google.com/github/catherinekamere-commits/AI-ML-Science-and-Engineering/blob/main/African_WASH_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

African Healthcare Facility Data: Assessing WASH Gaps Using Machine Learning


In [2]:
import pandas as pd

# The original parquet URL was incorrect. We can load the dataset directly from Hugging Face using pandas with the correct URL structure, or via the datasets library.
url = "https://huggingface.co/datasets/electricsheepafrica/africa-synth-wash-wash-in-healthcare-facilities-all/resolve/main/train.parquet"

try:
    df = pd.read_parquet(url)
except Exception:
    # Fallback to loading using Hugging Face datasets library if the direct link has complex partitioning
    !pip install -q datasets
    from datasets import load_dataset
    dataset = load_dataset("electricsheepafrica/africa-synth-wash-wash-in-healthcare-facilities-all")
    df = dataset["train"].to_pandas()

df.head()

README.md:   0%|          | 0.00/6.30k [00:00<?, ?B/s]

high_burden.csv:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

low_burden.csv:   0%|          | 0.00/939k [00:00<?, ?B/s]

moderate_burden.csv:   0%|          | 0.00/1.16M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/15000 [00:00<?, ? examples/s]

,facility_id,country,year,facility_type,location,has_water_source,water_source_type,water_reliability,has_sanitation,sanitation_facility_type,...,waste_disposal_method,sharps_disposal_method,has_cleaning_protocols,cleaning_supplies_available,cleaning_staff_trained,water_service_level,sanitation_service_level,hygiene_service_level,waste_service_level,cleaning_service_level
0,HCF000001,Rwanda,2020,clinic,urban,False,none,none,True,vip_latrines,...,open_burning,safety_boxes,False,False,False,no_service,limited,limited,limited,no_service
1,HCF000002,Uganda,2022,dispensary,rural,False,none,none,True,improved_pit_latrines,...,none,none,False,False,False,no_service,limited,limited,no_service,no_service
2,HCF000003,Mali,2021,health_center,urban,False,none,none,True,vip_latrines,...,none,none,True,True,True,no_service,limited,no_service,no_service,basic
3,HCF000004,Ghana,2018,health_center,urban,True,piped_municipal,mostly_available,True,flush_toilets,...,autoclave_incinerator,safety_boxes,False,False,False,limited,limited,no_service,basic,no_service
4,HCF000005,Rwanda,2023,clinic,urban,False,none,none,False,none,...,none,none,False,False,False,no_service,no_service,limited,no_service,no_service


In [3]:
df.shape

(15000, 30)

In [4]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 30 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   facility_id                  15000 non-null  object
 1   country                      15000 non-null  object
 2   year                         15000 non-null  int64 
 3   facility_type                15000 non-null  object
 4   location                     15000 non-null  object
 5   has_water_source             15000 non-null  bool  
 6   water_source_type            15000 non-null  object
 7   water_reliability            15000 non-null  object
 8   has_sanitation               15000 non-null  bool  
 9   sanitation_facility_type     15000 non-null  object
 10  num_toilets                  15000 non-null  int64 
 11  gender_separated_toilets     15000 non-null  bool  
 12  toilets_accessible_disabled  15000 non-null  bool  
 13  sanitation_usability         15

In [5]:
df.columns.tolist()

['facility_id',
 'country',
 'year',
 'facility_type',
 'location',
 'has_water_source',
 'water_source_type',
 'water_reliability',
 'has_sanitation',
 'sanitation_facility_type',
 'num_toilets',
 'gender_separated_toilets',
 'toilets_accessible_disabled',
 'sanitation_usability',
 'has_handwashing',
 'handwashing_locations',
 'soap_available',
 'alcohol_hand_rub_available',
 'has_waste_management',
 'waste_segregation',
 'waste_disposal_method',
 'sharps_disposal_method',
 'has_cleaning_protocols',
 'cleaning_supplies_available',
 'cleaning_staff_trained',
 'water_service_level',
 'sanitation_service_level',
 'hygiene_service_level',
 'waste_service_level',
 'cleaning_service_level']

In [6]:
print("Dataset shape:", df.shape)
print("\nData types:")
print(df.dtypes)

Dataset shape: (15000, 30)

Data types:
facility_id                    object
country                        object
year                            int64
facility_type                  object
location                       object
has_water_source                 bool
water_source_type              object
water_reliability              object
has_sanitation                   bool
sanitation_facility_type       object
num_toilets                     int64
gender_separated_toilets         bool
toilets_accessible_disabled      bool
sanitation_usability           object
has_handwashing                  bool
handwashing_locations          object
soap_available                   bool
alcohol_hand_rub_available       bool
has_waste_management             bool
waste_segregation              object
waste_disposal_method          object
sharps_disposal_method         object
has_cleaning_protocols           bool
cleaning_supplies_available      bool
cleaning_staff_trained           bool
water_serv

In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 30 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   facility_id                  15000 non-null  object
 1   country                      15000 non-null  object
 2   year                         15000 non-null  int64 
 3   facility_type                15000 non-null  object
 4   location                     15000 non-null  object
 5   has_water_source             15000 non-null  bool  
 6   water_source_type            15000 non-null  object
 7   water_reliability            15000 non-null  object
 8   has_sanitation               15000 non-null  bool  
 9   sanitation_facility_type     15000 non-null  object
 10  num_toilets                  15000 non-null  int64 
 11  gender_separated_toilets     15000 non-null  bool  
 12  toilets_accessible_disabled  15000 non-null  bool  
 13  sanitation_usability         15

In [8]:
# Check for missing values
missing = df.isnull().sum()

print("Missing values by column:")
print(missing[missing > 0])

Missing values by column:
Series([], dtype: int64)


In [9]:
# Check for duplicate records
print("Number of duplicate rows:", df.duplicated().sum())

Number of duplicate rows: 1


In [10]:
# Check unique values in categorical variables
for column in df.select_dtypes(include="object").columns:
    print(f"\n{column}:")
    print(df[column].unique()[:20])


facility_id:
['HCF000001' 'HCF000002' 'HCF000003' 'HCF000004' 'HCF000005' 'HCF000006'
 'HCF000007' 'HCF000008' 'HCF000009' 'HCF000010' 'HCF000011' 'HCF000012'
 'HCF000013' 'HCF000014' 'HCF000015' 'HCF000016' 'HCF000017' 'HCF000018'
 'HCF000019' 'HCF000020']

country:
['Rwanda' 'Uganda' 'Mali' 'Ghana' 'Ethiopia' 'Mozambique' 'Zambia'
 'Nigeria' 'DRC' 'Kenya' 'Tanzania' 'Niger' 'Senegal' 'Malawi']

facility_type:
['clinic' 'dispensary' 'health_center' 'hospital']

location:
['urban' 'rural' 'peri_urban']

water_source_type:
['none' 'piped_municipal' 'protected_well' 'borehole'
 'both_piped_borehole' 'rainwater']

water_reliability:
['none' 'mostly_available' 'frequently_unavailable' 'always_available']

sanitation_facility_type:
['vip_latrines' 'improved_pit_latrines' 'flush_toilets' 'none'
 'pit_latrines']

sanitation_usability:
['needs_repair' 'partially_functional' 'fully_functional' 'na']

handwashing_locations:
['most_points_of_care' 'some_points_of_care' 'none' 'all_points_of_care

Can we use information about a healthcare facility to predict its level of water service?



In [11]:
print(df["water_service_level"].value_counts())

water_service_level
no_service    6940
limited       4353
basic         3707
Name: count, dtype: int64


In [12]:
print(df["water_service_level"].value_counts(normalize=True))

water_service_level
no_service    0.462667
limited       0.290200
basic         0.247133
Name: proportion, dtype: float64


In [13]:
print(df["water_service_level"].value_counts())

water_service_level
no_service    6940
limited       4353
basic         3707
Name: count, dtype: int64


In [14]:
for column in [
    "country",
    "facility_type",
    "location",
    "water_source_type",
    "water_reliability",
    "has_water_source"
]:
    print(f"\n--- {column} ---")
    print(df[column].value_counts())


--- country ---
country
Mozambique    1121
Uganda        1114
Mali          1111
Niger         1100
Nigeria       1098
Ethiopia      1068
Kenya         1066
DRC           1065
Tanzania      1065
Malawi        1057
Senegal       1039
Zambia        1036
Ghana         1035
Rwanda        1025
Name: count, dtype: int64

--- facility_type ---
facility_type
health_center    5232
clinic           4539
dispensary       2968
hospital         2261
Name: count, dtype: int64

--- location ---
location
rural         5988
urban         5300
peri_urban    3712
Name: count, dtype: int64

--- water_source_type ---
water_source_type
none                   6940
piped_municipal        3064
borehole               2695
protected_well         1206
both_piped_borehole     747
rainwater               348
Name: count, dtype: int64

--- water_reliability ---
water_reliability
none                      6940
always_available          3707
mostly_available          3051
frequently_unavailable    1302
Name: count, d

In [15]:
pd.crosstab(
    df["has_water_source"],
    df["water_service_level"],
    normalize="index"
)

water_service_level,basic,limited,no_service
has_water_source,,,
False,0.000000,0.000000,1.0
True,0.459926,0.540074,0.0


In [16]:
pd.crosstab(
    df["water_reliability"],
    df["water_service_level"],
    normalize="index"
)

water_service_level,basic,limited,no_service
water_reliability,,,
always_available,1.0,0.0,0.0
frequently_unavailable,0.0,1.0,0.0
mostly_available,0.0,1.0,0.0
none,0.0,0.0,1.0


In [17]:
pd.crosstab(
    df["water_source_type"],
    df["water_service_level"],
    normalize="index"
)

water_service_level,basic,limited,no_service
water_source_type,,,
borehole,0.391095,0.608905,0.0
both_piped_borehole,0.546185,0.453815,0.0
none,0.000000,0.000000,1.0
piped_municipal,0.533290,0.466710,0.0
protected_well,0.404643,0.595357,0.0
rainwater,0.353448,0.646552,0.0


In [18]:
pd.crosstab(
    df["has_water_source"],
    df["water_service_level"],
    normalize="index"
)

water_service_level,basic,limited,no_service
has_water_source,,,
False,0.000000,0.000000,1.0
True,0.459926,0.540074,0.0


In [19]:
pd.crosstab(
    df["water_reliability"],
    df["water_service_level"],
    normalize="index"
)

water_service_level,basic,limited,no_service
water_reliability,,,
always_available,1.0,0.0,0.0
frequently_unavailable,0.0,1.0,0.0
mostly_available,0.0,1.0,0.0
none,0.0,0.0,1.0


In [20]:
# Check for missing values in our proposed features
features = [
    "country",
    "year",
    "facility_type",
    "location",
    "has_sanitation",
    "sanitation_facility_type",
    "num_toilets",
    "gender_separated_toilets",
    "toilets_accessible_disabled",
    "sanitation_usability",
    "has_handwashing",
    "handwashing_locations",
    "soap_available",
    "alcohol_hand_rub_available",
    "has_waste_management",
    "waste_segregation",
    "waste_disposal_method",
    "sharps_disposal_method",
    "has_cleaning_protocols",
    "cleaning_supplies_available",
    "cleaning_staff_trained"
]

df[features].isnull().sum()

,0
country,0
year,0
facility_type,0
location,0
has_sanitation,0
sanitation_facility_type,0
num_toilets,0
gender_separated_toilets,0
toilets_accessible_disabled,0
sanitation_usability,0


In [21]:
df[features].nunique().sort_values()

,0
gender_separated_toilets,2
has_sanitation,2
has_handwashing,2
has_waste_management,2
soap_available,2
alcohol_hand_rub_available,2
toilets_accessible_disabled,2
cleaning_supplies_available,2
has_cleaning_protocols,2
cleaning_staff_trained,2



# Assessing WASH Service Gaps in African Healthcare Facilities Using Machine Learning

## 1. Problem Statement

## 2. Intended AI/ML Outcome

## 3. Data Requirements

## 4. Dataset

## 5. Initial Data Exploration

## 6. Input Features and Target Variable

## 7. Proposed Machine Learning Workflow


Problem Statement

Access to reliable water, sanitation and hygiene (WASH) services is an important component of safe and effective healthcare delivery. However, the availability and quality of WASH services can vary across healthcare facilities.

This project aims to explore whether machine learning can be used to predict the level of water service available at healthcare facilities across Africa. The dataset contains information about healthcare facility characteristics, location, sanitation, hygiene, waste management and cleaning practices.

The machine learning problem will be treated as a **multiclass classification task**, where the model will predict one of three water service levels: **no service, limited service, or basic service**.


 Intended AI/ML Outcome

The intended outcome is to develop a machine learning classification model that can predict the **water service level** of a healthcare facility based on available facility and WASH-related characteristics.

The model will classify facilities into one of three categories:

* **No service**
* **Limited service**
* **Basic service**

The longer-term aim is to explore whether machine learning could support the identification of healthcare facilities that may require greater attention or investment in WASH infrastructure.



Data Requirements

To address the problem, data are required on characteristics that may be associated with WASH service availability in healthcare facilities.

The dataset should include information on:

* Country and geographic location
* Healthcare facility type
* Year of observation
* Sanitation facilities and usability
* Number and accessibility of toilets
* Handwashing facilities
* Availability of soap and alcohol hand rub
* Waste management and disposal
* Cleaning protocols, supplies and staff training
* Water service level, which will be used as the target variable

The selected dataset contains these types of information for healthcare facilities across 14 African countries.


Dataset

The project uses the **Africa Synthetic WASH in Healthcare Facilities** dataset obtained from Hugging Face.

The dataset contains **15,000 healthcare facility records** across 14 African countries.

The dataset includes variables covering:

* Facility characteristics
* Geographic location
* Water access and reliability
* Sanitation
* Hygiene
* Waste management
* Cleaning practices
* WASH service-level indicators

The dataset was imported directly into Google Colab for exploration and analysis.



Initial Data Exploration

Initial exploration was performed to understand the structure, completeness and distribution of the dataset.

The dataset contains **15,000 observations**. The target variable, `water_service_level`, contains three categories:

* **No service:** 6,940 facilities (46.3%)
* **Limited service:** 4,353 facilities (29.0%)
* **Basic service:** 3,707 facilities (24.7%)

The proposed predictor variables contained **no missing values**, indicating complete data for the variables considered during the initial assessment.

The dataset contains both categorical and numerical variables, including country, facility type, location, year and number of toilets, as well as several binary WASH indicators.


 Input Features and Target Variable

### Input Features

Potential input features for the machine learning model include:

* `country`
* `year`
* `facility_type`
* `location`
* `has_sanitation`
* `sanitation_facility_type`
* `num_toilets`
* `gender_separated_toilets`
* `toilets_accessible_disabled`
* `sanitation_usability`
* `has_handwashing`
* `handwashing_locations`
* `soap_available`
* `alcohol_hand_rub_available`
* `has_waste_management`
* `waste_segregation`
* `waste_disposal_method`
* `sharps_disposal_method`
* `has_cleaning_protocols`
* `cleaning_supplies_available`
* `cleaning_staff_trained`

### Output / Target

The target variable is:

**`water_service_level`**

The target contains three classes:

* `no_service`
* `limited`
* `basic`

### Data Governance Consideration

During the initial exploration, `water_source_type`, `water_reliability` and `has_water_source` were found to have a direct relationship with `water_service_level`. These variables may therefore introduce **target leakage** if used as predictors. They will require careful consideration and may be excluded from the final model to avoid artificially inflated model performance.


 Proposed Machine Learning Workflow

The proposed machine learning workflow is:

**Healthcare facility WASH dataset**
↓
**Data exploration and quality assessment**
↓
**Identify and address potential target leakage**
↓
**Select relevant input features and target variable**
↓
**Preprocess categorical and numerical data**
↓
**Split data into training and testing sets**
↓
**Train a multiclass classification model**
↓
**Evaluate model performance**
↓
**Interpret predictions and consider implications for WASH planning**

The initial modelling approach could compare a simple baseline classification model with a tree-based model such as Random Forest.
